opentelemetry-sdk/benches/metric.rs

```diff
@@ -53,9 +53,70 @@ impl MetricReader for SharedReader {
     }
 }
 
-fn bench_counter(view: Option<Box<dyn View>>) -> (Context, SharedReader, Counter<u64>) {
+/// Configure delta temporality for all [InstrumentKind]
+///
+/// [Temporality::Delta] will be used for all instrument kinds if this
+/// [TemporalitySelector] is used.
+#[derive(Clone, Default, Debug)]
+pub struct DeltaTemporalitySelector {
+    pub(crate) _private: (),
+}
+
+impl DeltaTemporalitySelector {
+    /// Create a new default temporality selector.
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl TemporalitySelector for DeltaTemporalitySelector {
+    fn temporality(&self, _kind: InstrumentKind) -> Temporality {
+        Temporality::Delta
+    }
+}
+
+// * Summary *
+
+// rustc 1.68.0 (2c8cc3432 2023-03-06)
+// cargo 1.68.0 (115f34552 2023-02-26), OS=Windows 11 Enterprise
+// Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz   2.59 GHz
+// 12 logical and 6 physical cores
+
+// Counter/AddNoAttrs      time:   [65.406 ns 65.535 ns 65.675 ns]
+// Counter/AddNoAttrsDelta time:   [65.553 ns 65.761 ns 65.981 ns]
+// Counter/AddOneAttr      time:   [341.55 ns 344.40 ns 347.58 ns]
+// Counter/AddOneAttrDelta time:   [340.11 ns 342.42 ns 344.89 ns]
+// Counter/AddThreeAttr    time:   [619.01 ns 624.16 ns 630.16 ns]
+// Counter/AddThreeAttrDelta
+//                         time:   [606.71 ns 611.45 ns 616.66 ns]
+// Counter/AddFiveAttr     time:   [3.7551 µs 3.7813 µs 3.8094 µs]
```

Comment:
Interesting, in Python, we never did anything specifically for perf involving number of attributes. Curious as to how you were able to achieve this specifically? Also I had to run the benchmark several times to be able to get increasing numbers like these (there were outliers every so often) so maybe we can run the test on the same machine used to run the dot net benchmarks to get a good baseline.

Reply:
From API standpoint: .NET has the above overloads for up to 3 attributes. 

counter.Add(long value, attribute)
counter.Add(long value, attribute1, attribute2)
counter.Add(long value, attribute1, attribute2, attribute3)

More than 3, we don't have more overloads :D, so one has to allocate an array and pass it. Allocating on heap is bad (.NET is garbage collected, so the goal is to avoid heap allocations as much as possible. 0 alloc is the goal for metrics!)
counter.Add(long value, attributes[]) --> passing an array (arrays are heap allocated).

From SDK, it keeps a per-thread Array for 1 attribute, 2 attributes, 3 attributes, .... 8 attributes. So up to 8 is alloc free (as it reuses the same array) from SDK. Anything more than 8, SDK has to allocate array, so perf drops.